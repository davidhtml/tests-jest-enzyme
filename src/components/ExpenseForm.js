import React from 'react';
import moment from "moment";
import {SingleDatePicker} from 'react-dates';
import 'react-dates/initialize'

class ExpenseForm extends React.Component {

    constructor(props) {
        super(props);

        this.state = {
            description: '',
            note: '',
            amount: '',
            createdAt: moment(),
            calendarFocused: null,
            errorMessage: null
        }
    }

    componentDidMount() {
        const {existingExpense} = this.props;
        if (existingExpense) {
            this.setState(() => {
                return {
                    description: existingExpense.description,
                    amount: (existingExpense.amount / 100).toString(),
                    note: existingExpense.note,
                    createdAt: moment(existingExpense.createdAt)
                }
            })
        }
    }

    onDescriptionChange = (e) => {
        const description = e.target.value;

        this.setState(() => ({ description: description}))
    }

    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({note: note}))
    }

    onAmountChange = (e) => {
        const amount = e.target.value;
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount: amount }))
        }
    }

    onDateChange = (createdAt) => {
        if (createdAt) {
            this.setState(() => ({createdAt: createdAt}))
        }

    }

    onFocusChange = (brazauskas) => {
        this.setState(() => ({ calendarFocused: brazauskas.focused }))
    }

    onSubmit = (e) => {
        e.preventDefault();
        if (!this.state.description || !this.state.amount) {
            let msg = 'Please provide the description and the amount';
            this.setState(() => ({errorMessage: msg}))
        } else {
            this.setState(() => ({errorMessage: null}))
            this.props.stateObjToParent({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10)*100,
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note

            });
        }
    }

    render() {
        return (

            <div>
                {this.state.errorMessage && <h4>{this.state.errorMessage}</h4>}
                <form onSubmit={this.onSubmit}>
                    <input
                        type="text"
                        placeholder="description"
                        autoFocus
                        value={this.state.description}
                        onChange={this.onDescriptionChange}
                    />
                    <input
                        type="text"
                        placeholder="amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                    />

                    <SingleDatePicker
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        id={this.state.description}
                        numberOfMonths={1}
                        isOutsideRange={() => false}
                    />


                    <textarea
                        placeholder="add a note for your expense"
                        value={this.state.note}
                        onChange={this.onNoteChange}
                        >
                    </textarea>
                    <button>Add expense</button>
                </form>
            </div>

        );
    }
}

export default ExpenseForm;
