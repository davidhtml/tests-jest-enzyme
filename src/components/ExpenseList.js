import React from 'react';
import {connect} from 'react-redux'
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

const ExpenseList = (props) => (
    <div className="content-container">
        <div className="list-header">
            <div className="show-for-mobile">Expenses</div>
            <div className="show-for-desktop">Expense</div>
            <div className="show-for-desktop">Amount</div>
        </div>
        <div className="list-body">
            {
                props.expenses.length === 0
                ? (
                    <div className="list-item--message">
                        You have no expenses
                    </div>
                )
                : (
                    props.expenses.map((expense) => {
                        return <ExpenseListItem key={expense.id} {...expense} />
                    })
                )
            }
        </div>
    </div>
);


const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    }
}


export { ExpenseList };
export default connect(mapStateToProps)(ExpenseList);;
