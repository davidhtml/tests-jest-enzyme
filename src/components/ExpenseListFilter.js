import React from 'react';
import {connect} from 'react-redux';
import {setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate} from '../actions/filters';
import {DateRangePicker} from 'react-dates';
import 'react-dates/initialize';

class ExpenseListFilters extends React.Component {

    state = {
        calendarFocused: null,
    }

    onDatesChange = (obj) => {
        this.props.setStartDate(obj.startDate);
        this.props.setEndDate(obj.endDate);
    }

    onFocusChange = (calendarFocused) => {
        console.log(calendarFocused)
        this.setState(() => ({ calendarFocused }))
    }

    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value);
    }

    onSortChange = (e) => {
        const option = e.target.value;
        if (option === 'date') {
            this.props.sortByDate();
        } else if (option === 'amount') {
            this.props.sortByAmount();
        }
    }

    render() {
        // console.log(this.props.filters.endDate)
        const {dispatch} = this.props;
        return (
            <div>
                <input
                    type="text"
                    value={this.props.filters.text}
                    onChange={this.onTextChange}/>
                <select
                    value={this.props.filters.sortBy}
                    onChange={this.onSortChange}>
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>
                <DateRangePicker
                    startDate={this.props.filters.startDate}
                    startDateId={"start"}
                    endDate={this.props.filters.endDate}
                    endDateId={"end"}
                    onDatesChange={this.onDatesChange}
                    focusedInput={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                    showClearDates={true}
                 />
            </div>
        );
    }
}

const mapStateToProps = (state) => {
    return {
        filters: state.filters,
    }
}

const mapDispatchToProps = (dispatch) => {
    return {
        setTextFilter: (text) => dispatch(setTextFilter(text)),
        sortByDate: () => dispatch(sortByDate()),
        sortByAmount: () => dispatch(sortByAmount()),
        setStartDate: (startDate) => dispatch(setStartDate(startDate)),
        setEndDate: (endDate) => dispatch(setEndDate(endDate))
    }
}

export { ExpenseListFilters };
export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);
