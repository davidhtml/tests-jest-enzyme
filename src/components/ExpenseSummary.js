import React from 'react';
import {connect} from 'react-redux';
import selectExpenses from '../selectors/expenses'
import expensesTotal from '../selectors/expenses-total';
import numeral from 'numeral'

class ExpenseSummary extends React.Component {

    renderExpenses = () => {
        const total = expensesTotal(this.props.expenses);
        return (
            <p>total of {this.props.expenses.length} expenses to pay {numeral(total/100).format('$0,0.00')}</p>
        )
    }

    render() {
        return (
            <div>
                <h4>Here will be how many visible expenses we do have and what is total amount</h4>
                {this.props.expenses.length > 0 && this.renderExpenses()}
            </div>
        );
    }
}

const mapStateToProps = state => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    }
}

export {ExpenseSummary}
export default connect(mapStateToProps)(ExpenseSummary);
