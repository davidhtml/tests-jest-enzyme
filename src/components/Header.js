import React from 'react';
import { NavLink } from 'react-router-dom';
import { connect } from 'react-redux';
import { startLogOut } from '../actions/auth';
const activeNav = {
    fontWeight: "bold",
    color: "#199"
}

const Header = (props) => (
    <header>
        <h1>Expensify.React App</h1>
        <p>Short description</p>
        <NavLink to="/dashboard" activeClassName="selected" activeStyle={activeNav}>Home</NavLink> <br/>
        <NavLink to="/create" activeClassName="selected" activeStyle={activeNav} exact>Create expense</NavLink> <br/>
        <NavLink to="/help" activeClassName="selected" activeStyle={activeNav} exact>Help</NavLink> <br/>
        <button onClick={props.logOut}>Log out</button>
    </header>
);

const mapDispatchToProps = (dispatch) => ({
    logOut: () => dispatch(startLogOut())
})

export { Header };
export default connect(null, mapDispatchToProps)(Header);
