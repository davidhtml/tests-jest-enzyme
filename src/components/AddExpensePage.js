import React from 'react';
import { connect } from 'react-redux';

import ExpenseForm from './ExpenseForm';
import { addExpense } from '../actions/expenses';


class AddExpensePage extends React.Component {

    onSubmit = (expense) => {
        this.props.addExpense(expense)
        this.props.history.push('/');
    }

    render() {
        return (
            <div>
                <h1>Add expense</h1>
                <ExpenseForm
                    stateObjToParent={this.onSubmit}
                 />
            </div>
        );
    }
}

const mapDispatchToProps = (dispatch) => {
    return {
        addExpense: (expense) => dispatch(addExpense(expense))
    };
};

export {AddExpensePage};

export default connect(undefined, mapDispatchToProps)(AddExpensePage);
