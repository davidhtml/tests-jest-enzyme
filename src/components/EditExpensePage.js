import React from 'react';
import {connect} from 'react-redux'
import { editExpense, removeExpense } from '../actions/expenses';
import ExpenseForm from './ExpenseForm';

//refactor to class bsaed compoennt
// create mapDispatchToProps

//should render addExpense page
// should handle editExpense using spies
//should handle removeExpense using spies


class EditExpensePage extends React.Component {

    onEditExpense = (expense) => {
        this.props.editExpense(this.props.expense.id, expense);
        this.props.history.push('/');
    }

    onRemove = () => {
        this.props.removeExpense({id: this.props.expense.id});
        this.props.history.push('/');
    }

    render() {
        return (
            <div>
                <ExpenseForm
                    stateObjToParent={this.onEditExpense}
                    existingExpense={this.props.expense}
                 />
                 <button
                     onClick={this.onRemove}
                >
                     Remove expense
                 </button>
            </div>
        );
    }

}



const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find((expense) => expense.id === props.match.params.id)
    };
};

const mapDispatchToProps = (dispatch) => {
    return {
        editExpense: (id, expense) => dispatch(editExpense(id, expense)),
        removeExpense: (id) => dispatch(removeExpense(id))
    };
};

export {EditExpensePage};
export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
