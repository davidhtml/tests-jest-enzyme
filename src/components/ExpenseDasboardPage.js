import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilter';
import ExpenseSummary from './ExpenseSummary';

const ExpenseDasboardPage = () => (
    <div>
        <ExpenseSummary />
        <ExpenseListFilters />
        <ExpenseList />
    </div>
);

export default ExpenseDasboardPage;
