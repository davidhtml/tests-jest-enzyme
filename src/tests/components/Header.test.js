import React from 'react';
// react-test-renderer
// import ShallowRenderer from 'react-test-renderer/shallow';
import { shallow } from 'enzyme';
// import toJSON from 'enzyme-to-json'
import { Header } from '../../components/Header'

test('should render Header correctly', () => {
    const wrapper = shallow(<Header logOut={() => {}}/>);
    expect(wrapper).toMatchSnapshot();
    //TODO enzyme example
    // expect(wrapper.find('p').text()).toBe("Short description");
    //TODO react-shallow-renderer example
    // const renderer = new ShallowRenderer();
    // renderer.render(<Header />);
    // expect(renderer.getRenderOutput()).toMatchSnapshot();
});


test('should call start logout on logOut', () => {
    const startLogOut = jest.fn();
    const wrapper = shallow(<Header logOut={startLogOut}/>);
    wrapper.find('button').simulate('click');
    expect(startLogOut).toHaveBeenCalled();
});
