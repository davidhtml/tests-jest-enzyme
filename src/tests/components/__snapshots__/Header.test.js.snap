// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Header correctly 1`] = `
<header>
  <h1>
    Expensify.React App
  </h1>
  <p>
    Short description
  </p>
  <NavLink
    activeClassName="selected"
    activeStyle={
      Object {
        "color": "#199",
        "fontWeight": "bold",
      }
    }
    aria-current="page"
    to="/dashboard"
  >
    Home
  </NavLink>
   
  <br />
  <NavLink
    activeClassName="selected"
    activeStyle={
      Object {
        "color": "#199",
        "fontWeight": "bold",
      }
    }
    aria-current="page"
    exact={true}
    to="/create"
  >
    Create expense
  </NavLink>
   
  <br />
  <NavLink
    activeClassName="selected"
    activeStyle={
      Object {
        "color": "#199",
        "fontWeight": "bold",
      }
    }
    aria-current="page"
    exact={true}
    to="/help"
  >
    Help
  </NavLink>
   
  <br />
  <button
    onClick={[Function]}
  >
    Log out
  </button>
</header>
`;
